File: neatformcore/src/main/java/com/nerdstone/neatformcore/form/json/JsonFormBuilder.java
package com.nerdstone.neatformcore.form.json;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import com.nerdstone.neatformcore.domain.model.form.NForm;
import com.nerdstone.neatformcore.domain.model.form.NFormContent;
import com.nerdstone.neatformcore.domain.view.FormBuilder;
import com.nerdstone.neatformcore.domain.view.RootView;
import com.nerdstone.neatformcore.rules.RulesFactory;
import com.nerdstone.neatformcore.rules.RulesFactory.RulesFileType;
import com.nerdstone.neatformcore.views.containers.VerticalRootView;
import com.nerdstone.neatformcore.views.handlers.ViewDispatcher;

import io.reactivex.CompletableObserver;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;

import static com.nerdstone.neatformcore.rx.RxHelper.getCompletableIoScheduler;

/***
 * @author Elly Nerdstone
 *
 */
public class JsonFormBuilder implements FormBuilder {

    private static final String TAG = JsonFormBuilder.class.getCanonicalName();
    private ViewGroup mainLayout;
    private ViewDispatcher viewDispatcher;
    private NForm form;
    private CompositeDisposable compositeDisposable;
    private RulesFactory rulesFactory;

    public JsonFormBuilder(ViewGroup mainLayout) {
        this.mainLayout = mainLayout;
        viewDispatcher = ViewDispatcher.getInstance();
        rulesFactory = RulesFactory.getInstance();
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public ViewGroup getMainLayout() {
        return this.mainLayout;
    }

    @Override
    public void setMainLayout(ViewGroup mainLayout) {
        this.mainLayout = mainLayout;
    }

    @Override
    public NForm getForm(String source) {
        if (form == null) {
            form = JsonFormParser.parseJson(source);
        }
        return form;
    }

    /***
     * @param context android context
     */
    @Override
    public void createFormViews(Context context) {
        if (form != null && form.getSteps() != null) {
            for (NFormContent formContent : form.getSteps()) {
                RootView rootView = new VerticalRootView(context);
                rootView.addChildren(formContent.getFields(), viewDispatcher);
                mainLayout.addView((View) rootView.initRootView());
            }
        }
    }

    @Override
    public void setViewDispatcher(ViewDispatcher viewDispatcher) {
        this.viewDispatcher = viewDispatcher;
    }

    @Override
    public void registerFormRules(Context context, RulesFileType rulesFileType) {
        rulesFactory.readRulesFromFile(context, form.getRulesFile(), rulesFileType)
                .compose(getCompletableIoScheduler()).subscribe(
                new CompletableObserver() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        compositeDisposable.add(d);
                    }

                    @Override
                    public void onComplete() {
                        Timber.tag(TAG).i("Completed reading rules from file successfully");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Timber.tag(TAG).e(e);
                    }
                });
    }

    @Override
    public void freeResources() {
        compositeDisposable.clear();
    }
}

File: neatformcore/src/main/java/com/nerdstone/neatformcore/form/json/JsonFormBuilder.kt
package com.nerdstone.neatformcore.form.json

import android.content.Context
import android.view.View
import android.view.ViewGroup
import com.nerdstone.neatformcore.datasource.AssetFile
import com.nerdstone.neatformcore.domain.builders.FormBuilder
import com.nerdstone.neatformcore.domain.model.NForm
import com.nerdstone.neatformcore.rules.RulesFactory
import com.nerdstone.neatformcore.rules.RulesFactory.RulesFileType
import com.nerdstone.neatformcore.utils.CoroutineContextProvider
import com.nerdstone.neatformcore.utils.SingleRunner
import com.nerdstone.neatformcore.utils.ViewUtils
import com.nerdstone.neatformcore.views.containers.VerticalRootView
import com.nerdstone.neatformcore.views.handlers.ViewDispatcher
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

/***
 * @author Elly Nerdstone
 */
class JsonFormBuilder(override var mainLayout: ViewGroup, override var fileSource: String) :
        FormBuilder {

    private val viewDispatcher: ViewDispatcher = ViewDispatcher.INSTANCE
    private val rulesFactory: RulesFactory = RulesFactory.INSTANCE
    private val rulesHandler = rulesFactory.rulesHandler
    private val singleRunner = SingleRunner()
    var coroutineContextProvider: CoroutineContextProvider
    var form: NForm? = null

    init {
        rulesHandler.formBuilder = this
        coroutineContextProvider = CoroutineContextProvider.Default()
    }

    override fun buildForm(viewList: List<View>?): FormBuilder {
        GlobalScope.launch(coroutineContextProvider.main) {
            if (form == null) {
                val async = async(coroutineContextProvider.default) {
                    singleRunner.afterPrevious {
                        parseJsonForm(fileSource)
                    }
                }
                form = async.await()
            }
            launch(coroutineContextProvider.io) {
                singleRunner.afterPrevious {
                    registerFormRules(mainLayout.context, RulesFileType.YAML)
                }
            }
            launch(coroutineContextProvider.main) {
                singleRunner.afterPrevious {
                    if (viewList == null)
                        createFormViews(mainLayout.context, arrayListOf())
                    else
                        createFormViews(mainLayout.context, viewList)
                }
            }
        }
        return this
    }

    private fun parseJsonForm(source: String): NForm? {
        return JsonFormParser.parseJson(
                AssetFile.readAssetFileAsString(
                        mainLayout.context,
                        source
                )
        )
    }

    /***
     * @param context android context
     */
    override fun createFormViews(context: Context, views: List<View>?) {
        if (form != null) {
            for ((index, formContent) in form!!.steps.withIndex()) {
                val rootView = VerticalRootView(context)
                val view = views?.getOrNull(index)
                if (view != null) {
                    ViewUtils.updateViews(view, formContent.fields, context, viewDispatcher)
                    rootView.addView(view)
                } else {
                    rootView.addChildren(formContent.fields, viewDispatcher)
                }
                mainLayout.addView(rootView.initRootView() as View)
            }
        }
    }


    override fun registerFormRules(context: Context, rulesFileType: RulesFileType) {
        form?.rulesFile?.also {
            rulesFactory.readRulesFromFile(context, it, rulesFileType)
        }
    }
}
